aws-lc-rs/src/key_wrap.rs

```diff
@@ -0,0 +1,354 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0 OR ISC
+
+//! Key Wrap Algorithm [NIST SP 800-38F](https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-38F.pdf)
+//!
+//! # Examples
+//! ```rust
+//! # use std::error::Error;
+//! # fn main() -> Result<(), Box<dyn Error>> {
+//! use aws_lc_rs::key_wrap::{KeyEncryptionKey, WrappingMode, AES_128};
+//!
+//! const KEY: &[u8] = &[
+//!     0xa8, 0xe0, 0x6d, 0xa6, 0x25, 0xa6, 0x5b, 0x25, 0xcf, 0x50, 0x30, 0x82, 0x68, 0x30, 0xb6,
+//!     0x61,
+//! ];
+//! const PLAINTEXT: &[u8] = &[0x43, 0xac, 0xff, 0x29, 0x31, 0x20, 0xdd, 0x5d];
+//!
+//! let kek = KeyEncryptionKey::new(&AES_128, KEY, WrappingMode::Padded)?;
+//!
+//! let mut output = vec![0u8; PLAINTEXT.len() + 15];
+//!
+//! let ciphertext = kek.wrap(PLAINTEXT, &mut output)?;
+//!
+//! let kek = KeyEncryptionKey::new(&AES_128, KEY, WrappingMode::Padded)?;
+//!
+//! let mut output = vec![0u8; ciphertext.len()];
+//!
+//! let plaintext = kek.unwrap(&*ciphertext, &mut output)?;
+//!
+//! assert_eq!(PLAINTEXT, plaintext);
+//!
+//! Ok(())
+//! # }
+//! ```
+use std::{fmt::Debug, mem::MaybeUninit, ptr::null};
+
+use aws_lc::{
+    AES_set_decrypt_key, AES_set_encrypt_key, AES_unwrap_key, AES_unwrap_key_padded, AES_wrap_key,
+    AES_wrap_key_padded, AES_KEY,
+};
+
+use crate::{error::Unspecified, fips::indicator_check};
+
+mod tests;
+
+/// The Key Wrapping Algorithm
+pub struct Algorithm {
+    id: AlgorithmId,
+    key_len: usize,
+}
+
+impl Algorithm {
+    /// Returns the algorithm identifier.
+    #[inline]
+    #[must_use]
+    pub fn id(&self) -> AlgorithmId {
+        self.id
+    }
+
+    /// Returns the algorithm key length.
+    #[inline]
+    #[must_use]
+    fn key_len(&self) -> usize {
+        self.key_len
+    }
+}
+
+impl Debug for Algorithm {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        Debug::fmt(&self.id, f)
+    }
+}
+
+/// The Key Wrapping Algorithm Identifier
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+pub enum AlgorithmId {
+    /// AES-128 Key Wrap
+    Aes128,
+
+    /// AES-256 Key Wrap
+    Aes256,
+}
```

Comment:
Since there are [other algorithms for key wrapping](https://en.wikipedia.org/wiki/Key_wrap), should we designate these as `AESKW_Aes128` and `AESKW_Aes256`?

Do we anticipate ever needing to expand this API to support other forms of "key wrapping"?